client/src/main/scala/drt/client/components/TerminalStaffingV2.scala

```diff
@@ -0,0 +1,266 @@
+import scala.scalajs.js
+import scala.scalajs.js.annotation.JSImport
+import scala.util.{Failure, Success, Try}
+
+object HotTable {
+
+  val log: Logger = LoggerFactory.getLogger("TerminalStaffing")
+
+  @JSImport("react-handsontable", JSImport.Default)
+  @js.native
+  object RawComponent extends js.Object
+
+  @js.native
+  trait Props extends js.Object {
+    var settings: js.Dictionary[js.Any] = js.native
+  }
+
+  def props(
+             data: Seq[Seq[AnyVal]],
+             colHeadings: Seq[String],
+             rowHeadings: Seq[String],
+             changeCallback: (Int, Int, Int) => Unit,
+             colWidths: String = "2em"
+           ): Props = {
+    import js.JSConverters._
+    val p = (new js.Object).asInstanceOf[Props]
+    val afterChange = (changes: js.Array[js.Array[Any]], source: String) => {
+      val maybeArray = Option(changes)
+      maybeArray.foreach(
+        c => {
+          c.toList.foreach(change =>
+            (change(0), change(1), change(3)) match {
+              case (row: Int, col: Int, value: String) =>
+                val tryValue = Try(Integer.parseInt(value)) match {
+                  case Success(v) =>
+                    changeCallback(row, col, v)
+                  case Failure(f) =>
+                    log.warn(s"Couldn't parse $value to an Integer $f")
+                }
+
+              case (row: Int, col: Int, value: Int) =>
+                changeCallback(row, col, value)
+              case other =>
+                log.error(s"couldn't match $other")
+            }
+          )
+        })
+      if (maybeArray.isEmpty) {
+        log.info(s"Called change function with no values")
+      }
+    }
+
+    p.settings = js.Dictionary(
+      "data" -> data.map(_.toJSArray).toJSArray,
+      "rowHeaders" -> rowHeadings.toJSArray,
+      "colHeaders" -> colHeadings.toJSArray,
+      "afterChange" -> afterChange,
+      "colWidth" -> colWidths
+    )
+    p
+  }
+
+  val component = JsComponent[Props, Children.None, Null](RawComponent)
+}
+
+object TerminalStaffingV2 {
+
+  case class TimeSlotDay(timeSlot: Int, day: Int) {
+    def key = s"$timeSlot-$day"
+  }
+
+  case class State(
+                    timeSlots: Seq[Seq[Int]],
+                    colHeadings: Seq[String],
+                    rowHeadings: Seq[String],
+                    changes: Map[String, Int]
+                  )
+
+  val log: Logger = LoggerFactory.getLogger(getClass.getName)
+
+  case class Props(
+                    rawShiftString: String,
+                    terminalPageTab: TerminalPageTabLoc,
+                    router: RouterCtl[Loc]
+                  ) {
+    def timeSlotMinutes = Try(terminalPageTab.subMode.toInt).toOption.getOrElse(60)
+  }
+
+  def staffToStaffTimeSlotsForMonth(month: SDateLike, staff: Seq[Seq[Int]], terminal: String, slotMinutes: Int): StaffTimeSlotsForTerminalMonth = {
+    StaffTimeSlotsForTerminalMonth(month.millisSinceEpoch, terminal, staff.zipWithIndex.flatMap {
```

Comment:
take the staff flatmap stuff out into its own function to make the case class construction clearer?